katcp/resource_client.py

```diff
@@ -881,33 +881,37 @@ def list_sensors(self, filter="", strategy=False, status="",
     def set_sensor_strategy(self, sensor_name, strategy_and_parms):
         sensor_name = resource.escape_name(sensor_name)
         sensor_obj = getattr(self.sensor, sensor_name, None)
-        for child_name in dict.keys(self.children):
+        if sensor_obj:
+            # The sensor exists, so set the strategy.
+            child_name = sensor_obj.parent_name
             child = self.children[child_name]
+            # Handle aggregate sensors that are not prefixed with "parent_name_"
             if sensor_name.startswith("agg_"):
-                if sensor_obj and (child.name == sensor_obj.parent_name):
-                    # Handle aggregate sensors that are not prefixed with "parent_name_"
-                    yield child.set_sensor_strategy(sensor_name, strategy_and_parms)
+                yield child.set_sensor_strategy(sensor_name, strategy_and_parms)
             else:
                 # Get the child_sensor_name without the parent_name prefix
                 prefix = child_name + '_'
                 child_sensor_name = sensor_name[len(prefix):]
                 yield child.set_sensor_strategy(child_sensor_name, strategy_and_parms)
+        # Otherwise, depend on self._add_sensors() to handle it when the sensor appears
 
     @steal_docstring_from(resource.KATCPResource.set_sensor_listener)
     def set_sensor_listener(self, sensor_name, listener):
         sensor_name = resource.escape_name(sensor_name)
         sensor_obj = getattr(self.sensor, sensor_name, None)
-        for child_name in dict.keys(self.children):
+        if sensor_obj:
+            # The sensor exists, so register the listener.
+            child_name = sensor_obj.parent_name
```

Comment:
Yes - the parent of sensor_obj is a child of the container.
Will see if there is some rework that can be done on the naming.